ccm/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+use aead::{AeadInPlace, Error, Key, NewAead, Nonce, Tag};
+use block_cipher::{Block, BlockCipher, NewBlockCipher};
+use core::marker::PhantomData;
+use subtle::ConstantTimeEq;
+
+mod traits;
+
+use traits::{NonceSize, TagSize};
+
+/// CCM instance generic over an underlying block cipher.
+///
+/// Type parameters:
+/// - `C`: block cipher.
+/// - `M`: size of MAC tag, valid values:
+/// `U4`, `U6`, `U8`, `U10`, `U12`, `U14`, `U16`.
+/// - `N`: size of nonce, valid values:
+/// `U7`, `U8`, `U9`, `U10`, `U11`, `U12`, `U13`.
+pub struct Ccm<C, M, N>
+where
+    C: BlockCipher<BlockSize = U16>,
+    C::ParBlocks: ArrayLength<Block<C>>,
+    M: ArrayLength<u8> + TagSize,
+    N: ArrayLength<u8> + NonceSize,
+{
+    cipher: C,
+    _tag_size: PhantomData<M>,
+    _nonce_size: PhantomData<N>,
+}
+
+impl<C, M, N> Ccm<C, M, N>
+where
+    C: BlockCipher<BlockSize = U16>,
+    C::ParBlocks: ArrayLength<Block<C>>,
+    M: ArrayLength<u8> + TagSize,
+    N: ArrayLength<u8> + NonceSize,
+{
+    fn from_cipher(cipher: C) -> Self {
+        Self {
+            cipher,
+            _tag_size: Default::default(),
+            _nonce_size: Default::default(),
+        }
+    }
+
+    fn gen_enc_block(block: &mut Block<C>, nonce: &Nonce<N>, i: usize) {
+        block[0] = N::get_l() - 1;
+        let n = 1 + N::to_usize();
+        block[1..n].copy_from_slice(nonce);
+        let b = &mut block[n..];
+        let arr = i.to_be_bytes();
+        b.copy_from_slice(&arr[arr.len() - b.len()..])
+    }
+
+    fn calc_mac(
+        &self,
+        nonce: &Nonce<N>,
+        adata: &[u8],
+        buffer: &[u8],
+    ) -> Result<Tag<C::BlockSize>, Error> {
+        let bs = C::BlockSize::to_usize();
+        let is_ad = !adata.is_empty();
+        let l = N::get_l();
+        let flags = 64 * (is_ad as u8) + 8 * M::get_m_tick() + (l - 1);
+
+        if buffer.len() >= (1 << (8 * l)) {
+            return Err(Error);
+        }
+        let l_arr = buffer.len().to_be_bytes();
+
+        let mut b0 = Block::<C>::default();
+        b0[0] = flags;
+        let n = 1 + N::to_usize();
+        b0[1..n].copy_from_slice(&nonce);
+        let q = l_arr.len() - l as usize;
+        b0[n..].copy_from_slice(&l_arr[q..]);
+
+        let la = adata.len();
+        let la_arr = la.to_be_bytes();
+        let b1 = if la == 0 {
+            None
+        } else if la < (1 << 16) - (1 << 8) {
+            let mut b = Block::<C>::default();
+            b[..2].copy_from_slice(&la_arr[la_arr.len() - 2..]);
+            Some((b, 2))
+        } else if la <= core::u32::MAX as usize {
+            let mut b = Block::<C>::default();
+            b[0] = 0xFF;
+            b[1] = 0xEE;
+            b[2..6].copy_from_slice(&la_arr[la_arr.len() - 4..]);
+            Some((b, 6))
+        } else {
+            let mut b = Block::<C>::default();
+            b[0] = 0xFF;
+            b[1] = 0xFF;
+            b[2..10].copy_from_slice(&la_arr[la_arr.len() - 8..]);
+            Some((b, 10))
```

Comment:
This branch is not covered by tests, but to fix that, we would have to test AD bigger than 4 GiB. With `aesni` as a backend in release mode it takes several seconds, but with `aes-soft` it takes more than a minute.